pkg/networking/backend_sg_provider.go

```diff
@@ -0,0 +1,245 @@
+	"github.com/pkg/errors"
+	networking "k8s.io/api/networking/v1beta1"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/aws/services"
+	"sigs.k8s.io/aws-load-balancer-controller/pkg/runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+	"sync"
+	"time"
+)
+
+const (
+	resourceTypeSecurityGroup = "security-group"
+	tagKeyK8sCluster          = "elbv2.k8s.aws/cluster"
+	tagKeySGType              = "elbv2.k8s.aws/type"
+	tagValueBackend           = "backend"
+
+	explicitGroupFinalizerPrefix = "group.ingress.k8s.aws/"
+	implicitGroupFinalizer       = "ingress.k8s.aws/resources"
+
+	sgDescription = "[k8s] Shared Backend SecurityGroup for LoadBalancer"
+)
+
+// BackendSGProvider is responsible for providing backend security groups
+type BackendSGProvider interface {
+	// Get returns the backend security groups to use
+	Get(ctx context.Context) ([]string, error)
+	// Release cleans up the auto-generated backend SG if necessary
+	Release(ctx context.Context) error
+}
+
+// NewBackendSGProvider constructs a new  defaultBackendSGProvider
+func NewBackendSGProvider(clusterName string, backendSGEnabled bool, backendSGs []string, vpcID string,
+	ec2Client services.EC2, k8sClient client.Client, logger logr.Logger) *defaultBackendSGProvider {
+	var sanitizedBackendSGs []string
+	for _, sg := range backendSGs {
+		sanitizedBackendSGs = append(sanitizedBackendSGs, strings.TrimSpace(sg))
+	}
+	return &defaultBackendSGProvider{
+		vpcID:            vpcID,
+		clusterName:      clusterName,
+		backendSGEnabled: backendSGEnabled,
+		backendSGs:       sanitizedBackendSGs,
+		ec2Client:        ec2Client,
+		k8sClient:        k8sClient,
+		logger:           logger,
+		mutex:            sync.Mutex{},
+	}
+}
+
+var _ BackendSGProvider = &defaultBackendSGProvider{}
+
+type defaultBackendSGProvider struct {
+	vpcID       string
+	clusterName string
+	mutex       sync.Mutex
+
+	backendSGEnabled bool
+	backendSGs       []string
+	autoGeneratedSG  string
+	ec2Client        services.EC2
+	k8sClient        client.Client
+	logger           logr.Logger
+}
+
+func (p *defaultBackendSGProvider) Get(ctx context.Context) ([]string, error) {
+	if !p.backendSGEnabled {
+		return nil, nil
+	}
+	if len(p.backendSGs) > 0 {
+		return p.backendSGs, nil
+	}
+	// Auto generate Backend Security group, and return the id
+	if err := p.allocateBackendSG(ctx); err != nil {
+		p.logger.Error(err, "Failed to auto-create backend SG")
+		return nil, err
+	}
+	return []string{p.autoGeneratedSG}, nil
+}
+
+func (p *defaultBackendSGProvider) Release(ctx context.Context) error {
+	if !p.backendSGEnabled || len(p.backendSGs) > 0 {
+		return nil
+	}
+	if required, err := p.isBackendSGRequired(ctx); required || err != nil {
+		return err
+	}
+	return p.releaseSG(ctx)
+}
+
+func (p *defaultBackendSGProvider) allocateBackendSG(ctx context.Context) error {
+	p.mutex.Lock()
+	defer p.mutex.Unlock()
+
+	if len(p.autoGeneratedSG) > 0 {
+		return nil
+	}
+
+	sgName := p.getBackendSGName()
```

Comment:
let's use resource tags instead of name to identity existing resources in cluster. 

Name pattern might need to change given our past experience. (e.g. deprecate a unsafe hash algorithm)

Reply:
I was trying to avoid getting the tracking provider in this module - didn't remember if I ran into dependency issues earlier.

Reply:
done